qiskit_experiments/curve_analysis/grouped_curve_analysis.py

```diff
@@ -0,0 +1,324 @@
+    r"""Curve analysis with multiple independent groups.
+
+    The :class:`.MultiGroupCurveAnalysis` provides a fitting framework for the case
+
+    .. math::
+
+        \Theta_{i, \mbox{opt}} = \arg\min_{\Theta_{i, \mbox{fit}}} (F(X_i, \Theta_i)-Y_i)^2,
+
+    where :math:`F` is a common fit model for multiple independent dataset :math:`(X_i, Y_i)`.
+    The fit generates multiple fit outcomes :math:`\Theta_{i, \mbox{opt}}` for each dataset.
+
+    Each experiment circuit must have metadata of data sorting keys for model mapping and
+    group mapping. For example, if the experiment has two experiment conditions "A" and "B"
+    while fitting the outcomes with three models "X", "Y", "Z", then the metadata must consist
+    of the condition and model, e.g. ``{"condition": "A", "model": "X"}``.
+
+    In this example, "condition" should be defined in the ``group_data_sort_key``
+    of the :class:`.MultiGroupCurveAnalysis`, and "model" should be defined in
+    the ``data_sort_key`` for each LMFIT model.
+
+    .. code-block:: python
+
+        import lmfit
+        from qiskit_experiments.curve_analysis import MultiGroupCurveAnalysis
+
+        analysis = MultiGroupCurveAnalysis(
+            groups=["groupA", "groupB"],
+            group_data_sort_key=[{"condition": "A"}, {"condition": "B"}],
+            models=[
+                lmfit.Model(fit_func_x, name="curve_X", data_sort_key={"model": "X"}),
+                lmfit.Model(fit_func_y, name="curve_Y", data_sort_key={"model": "Y"}),
+                lmfit.Model(fit_func_z, name="curve_Z", data_sort_key={"model": "Z"}),
+            ]
+        )
+
+    In above setting, each fit curve will appear in the output figure with unique name
+    in the form of ``{group name}_{model name}``, e.g. ``groupA_curve_X``.
+
+    A subclass can implement the following extra method.
+
+    .. rubric:: _create_composite_analysis_results
+
+    This method computes new quantities by taking all fit outcomes and
+    store the new values in the analysis results.
+    By default, no extra quantity is computed.
+
+    """
+
+    def __init__(
+        self,
+        groups: List[str],
+        group_data_sort_key: List[Dict[str, Any]],
+        models: Optional[List[lmfit.Model]] = None,
+    ):
+        """Create new analysis.
+
+        Args:
+            groups: List of group names.
+            group_data_sort_key: List of sort key dictionary for each group.
+            models: List of LMFIT ``Model`` class to define fitting functions and
+                parameters. Provided models are shared among groups.
+
+        Raises:
+            AnalysisError: When groups and group_data_sort_key don't match.
+        """
+        super().__init__()
+
+        if len(groups) != len(group_data_sort_key):
+            raise AnalysisError(
+                "Number of groups and data sort key doesn't match. "
+                "Data sort key must be provided for each group."
+            )
```

Comment:
Is there some redundancy between `groups` and `group_data_sort_key`? This suggests so. Do we need to specify both, i.e., could we somehow do this with one variable?

Reply:
Good point. Actually there is some duplication, but I feel merging them makes data structure confusing, i.e. `Dict[str, Dict[str, Any]]` e.g. `{"groupA": {"setting": "A"}, "groupB": {"setting": "B"}}`. If this looks okey perhaps we can update the signature.